common/context/src/main/java/io/helidon/common/context/Contexts.java

```diff
@@ -0,0 +1,129 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.helidon.common.context;
+
+import java.util.Optional;
+import java.util.Stack;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+
+/**
+ * Support for handling {@link io.helidon.common.context.Context} across thread boundaries.
+ */
+public final class Contexts {
+    static final ThreadLocal<Stack<Context>> REGISTRY = ThreadLocal.withInitial(Stack::new);
+
+    private Contexts() {
+    }
+
+    static void clear() {
+        REGISTRY.get().clear();
+    }
+
+    static void push(Context context) {
+        REGISTRY.get().push(context);
+    }
+
+    static Context pop() {
+        return REGISTRY.get().pop();
+    }
+
+    /**
+     * Get context registry associated with current thread.
+     *
+     * @return context that is associated with current thread or an empty context if none is
+     */
+    public static Optional<Context> context() {
+        Stack<Context> contextStack = REGISTRY.get();
+
+        if (contextStack.isEmpty()) {
+            return Optional.empty();
+        }
+
+        return Optional.ofNullable(contextStack.peek());
+    }
+
+    /**
+     * Wrap an executor service to correctly propagate context to its threads.
+     *
+     * @param toWrap executor service
+     * @return a new executor service wrapping the provided one
+     */
+    public static ExecutorService wrap(ExecutorService toWrap) {
+        if (toWrap instanceof ContextAwareExecutorImpl) {
+            return toWrap;
+        }
+        if (toWrap instanceof ContextAwareScheduledExecutorImpl) {
+            return toWrap;
+        }
+        return new ContextAwareExecutorImpl(toWrap);
+    }
+
+    /**
+     * Wrap a scheduled executor service to correctly propagate context to its threads.
+     * Note that all scheduled methods are going to run in context of the thread scheduling the tasks.
+     *
+     * @param toWrap executor service
+     * @return a new executor service wrapping the provided one
+     */
+    public static ScheduledExecutorService wrap(ScheduledExecutorService toWrap) {
+        if (toWrap instanceof ContextAwareScheduledExecutorImpl) {
+            return toWrap;
+        }
+        return new ContextAwareScheduledExecutorImpl(toWrap);
+    }
+
+    /**
+     * Run the runnable in the provided context.
+     * The runnable can use {@link #context()} to retrieve the context.
+     *
+     * @param context  context to run in
+     * @param runnable runnable to execute in context
+     */
+    public static void inContext(Context context, Runnable runnable) {
```

Comment:
When seen in use, it isn't immediately apparent that `Context.inContext(...)` means `run`. It would be clearer if this method was called `runInContext` or `invokeInContext`.

You could also consider moving this to a non-static method on `Context`, so the call becomes e.g.`context.run(runnable)` or `context.invoke(callable)`.

Reply:
@spericas what do you think about this? I wanted to have the static methods grouped into the `Contexts` class.
You are right this could be an instance method on context itself. 
In that case, I would have to implement is as a default method. And if anybody ever decides to implement the `Context` interface and they also implement the `run` or `invoke` method, this may break our handling of contexts (as the method `push` and `pop` is package local and should not be a part of our public API.
Not sure whether we want to open the API in this way.

Reply:
Yes, there are definitely tradeoffs making them non-static. I'm OK keeping them static but still think they should be renamed.

Reply:
Renamed

Reply:
@tomas-langer I also have a small preference for the non-static variant, but at this point I feel I need to experiment with all this in real use cases (like FT) before I can provide more detailed feedback, so I'm OK with how it looks now (method renaming helps).